slowmovie.py

```diff
@@ -10,205 +10,349 @@
+                print(f'Reading logfile failed, caught following error: {e}. Starting at beginning of video.')
+                currentFrame = 0
+    else:
+        currentFrame = 0
+
+# Set lastVideo so that first time through the loop, we'll print "Playing x"
+lastVideo = None
+
+while True:
+    # Print a message when starting a new video
+    if lastVideo != currentVideo:
+        print(f"Playing '{currentVideo}'")
+        print(f'Video info: {videoInfo["frame_count"]} frames, {videoInfo["fps"]:.3f}fps, duration: {videoInfo["duration"]}s')
+        print(f'This video will take {estimate_runtime(args.delay, args.increment, videoInfo["duration"], videoInfo["fps"], "a")} to play.')
+        lastVideo = currentVideo
+
+    # Note the time when starting to display so we can adjust for how long it takes later
+    displayStartTime = time.perf_counter()
 
-    log = open(logdir + '%s<progress'%currentVideo, 'w')
-    log.write(str(currentPosition))
-    log.close() 
+    epd.init()
 
+    if args.random_frames:
+        currentFrame = random.randint(0,videoInfo['frame_count'])
 
-    f = open('nowPlaying', 'w')
-    f.write(currentVideo)
-    f.close() 
-    
+    msTimecode = f'{currentFrame*videoInfo["frame_time"]}ms'
 
-#     epd.sleep()
-    time.sleep(frameDelay)
-    epd.init()
+    # Use ffmpeg to extract a frame from the movie, crop it, letterbox it, and put it in memory as frame.bmp
+    generate_frame(videoFilepath, '/dev/shm/frame.bmp', msTimecode)
 
+    # Open frame.bmp in PIL
+    pil_im = Image.open('/dev/shm/frame.bmp')
 
+    # Adjust contrast if specified
+    if args.contrast != 1:
+        enhancer = ImageEnhance.Contrast(pil_im)
+        pil_im = enhancer.enhance(args.contrast)
 
+    # Dither the image into a 1 bit bitmap (Just zeros and ones)
+    pil_im = pil_im.convert(mode = '1',dither = Image.FLOYDSTEINBERG)
+
+    # Display the image
+    print(f'Displaying frame {int(currentFrame)} of {currentVideo} ({(currentFrame/videoInfo["frame_count"])*100:.1f}%)')
+    epd.display(epd.getbuffer(pil_im))
 
-epd.sleep()
-    
-epd7in5.epdconfig.module_exit()
-exit()
+    # Increment the position
+    if not args.random_frames:
+        currentFrame += args.increment
+        # If it's the end of the video
+        if currentFrame > videoInfo['frame_count']:
+            if not args.loop:
+                if args.random_file:
+                    # Pick a new random video
+                    currentVideo = random.choice(videos)
+                else:
+                    # Inrcrement fileIndex or wrap around
+                    fileIndex += 1
+                    if fileIndex >= len(videos):
+                        fileIndex = 0
+
+                    # Update currently playing video to be the incremented one
+                    currentVideo = videos[fileIndex]
+
+                # Note the new video we picked in nowPlaying file
+                with open('nowPlaying', 'w') as file:
```

Comment:
should this file be a global constant ```NOW_PLAYING="nowPlaying"``` and then use a function to read/write the files? Just cleaner but not necessary. One mistype on the filename somewhere would mess things up.

Reply:
Upon review maybe I'm just being picky here :) It's fine the way it is for now. If something else requires touching the file IO code in the future it can be dealt with then. There will always be more PRs. @qubist go ahead and resolve this unless you plan on doing something with it.

Reply:
5bebbc7 Implemented. Oops I named that commit wrong. Should say "...writing to nowPlaying file". Reading from it is not refactored, but it's only done in one spot so I'm leaving it

Reply:
Whoops saw your update too late :)

Reply:
@robweber we can rollback that commit if what I did is less desirable than the way it was. Such a little thing it's agonizing to decide XD